nixos/apparmor: policy activation tristate and policy path support

LordGrimmauld opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

AppArmor profiles are files written to `/etc/apparmor.d/<name>`. Previously, profiles were defined with a string and written to a store entry, then linked into the etc path. However, this makes little sense if one wants to use profiles that are already files in the nix store, e.g. when using rules from `roddhjav-apparmor-rules` (which is absolutely broken currently, but that is a different topic; PR for that coming eventually). This change in the module exposes an option `security.apparmor.policies.<name>.path` with which a path to a profile can be directly symlinked into the apparmor directory, without first reading and rewriting that file into nix store again. This should boost nixos eval times on systems using prebuilt apparmor rules and makes prebuilding/caching apparmor rulesets via nix possible.

AppArmor profiles always fall within a tristate of `enable`, `disable`, `complain`. It makes no sense to have two boolean options instead of a neat enum config for the tristate. Therefore, i added that tristate, removing the two bools in the process.

The apparmor nixos test breaks often. While testing this PR, i noticed the test failing on master/nixos-unstable too. Therefore, i included a fix to the nixos test.

Yes, the two removed bools are a breaking change. However, the path option addition is inert, it does not break definitions of apparmor policies defined via string.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

LordGrimmauld committed:
nixos/apparmor: Format

LordGrimmauld committed:
nixos/apparmor: profile activation tristate and profile path support

LordGrimmauld commented on nixos/modules/security/apparmor.nix:
```diff
@@ -1,23 +1,32 @@
+      "security"
+      "apparmor"
+      "confineSUIDApplications"
+    ] "Please use the new options: `security.apparmor.policies.<policy>.enable'.")
```
Completely right, i missed that one. It is doc, so everything worked, but it should say state of course. That was the only miss of that rename afaict. Fixed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/apparmor-default-profiles/16780/8

symphorien commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +51,55 @@ in
+                  ];
+                  # should enforce really be the default?
+                  # the docs state that this should only be used once one is REALLY sure nothing's gonna break
+                  default = "enforce";
```
alternatively if you don't set a default users will have to specify a value explicitly for each policy

LordGrimmauld replied:
I put `default = "enforce";` because this was default before this PR.
The problem is, setting a different default would mean behavior would *silently* change in the background, i do not believe this is a good idea.
Not setting a default would mean existing configs relying on an old default would at least *fail loud*. However, it stands to reason this would actively break even more configs that would otherwise just silently use the new implementation with the old effective behavior.

symphorien replied:
makes sense

symphorien commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +51,55 @@ in
+                    assert (
+                      lib.assertMsg (
+                        (p != null && !options.profile.isDefined) || (p == null && options.profile.isDefined)
+                      ) "`security.apparmor.policies.\"${name}\"` must define exactly one of either path or profile."
```
alternatively you could use the assertion mechanism of nixos https://nixos.org/manual/nixos/unstable/#sec-assertions

LordGrimmauld replied:
my understanding is assertions do really not like submodules. Am i mistaken here?

symphorien replied:
honestly I don't know

getchoo replied:
> Am i mistaken here?

This is correct, but I think there is a better solution here

Since assertions don't really exist in submodule, we can instead create these assertions from the top-level context by mapping over each policy (we can simplify this assertion with an XOR operator)

```nix
{
  config = {
    assertions = lib.mapAttrsToList (policyName: policyCfg: {
      assertion = lib.xor policyCfg.profile.isDefined (policyCfg.path != null);
      message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
    }) cfg.policies;
  };
}
```

LordGrimmauld replied:
Alright, i implemented the assert on the global module level now. You have a point, and i do actually like this way better.

symphorien commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +51,55 @@ in
+                };
+
+                profile = lib.mkOption {
+                  description = "The policy of the profile. Incompatible with path.";
```
```suggestion
                  description = "The policy of the profile, i.e. the desired content of `/etc/apparmor.d/<name>.conf`. Incompatible with path.";
```

I'm not very familiar with apparmor but the current formulation is really obscure to me.

symphorien commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +51,55 @@ in
+            {
+              options = {
+                state = lib.mkOption {
+                  description = "The state of the profile as applied to the system by nix";
```
```suggestion
                  description = "How strictly this policy should be enforced";
```

nix the package manager has little to do with all this

symphorien reviewed (commented):
do you think you could test your options in the nixos test?

LordGrimmauld committed:
nixos/tests/apparmor.nix: Format

LordGrimmauld committed:
nixos/tests/apparmor.nix: fix expected.rules

LordGrimmauld committed:
nixos/tests/apparmor: move to folder, refactor, improve coverage

- nixfmt on apparmor test
- move apparmor test to nixos/tests/apparmor directory
- expected profile contents are now generated in its own file to make the test file less confusing and hard to maintain
- enforce/complain is now being tested via diff of expected against aa-status
- path is now tested against diff+file checking symlink target of /etc/static/apparmor.d/<name>
- profile is now checked by diff of /etc/static/apparmor.d/<name> against original string added in nix config
- test still successfully passes
- added test for confined hello to succeed
- added test for confined hexdump on denied path to fail

symphorien commented:
Thank you for the improved nixos test. Do you think you could add a profile in enforce mode for a third tool (say, hexdump) and check that it really gets permission denied/crashes when trying to access files beyond what its profile allows ? That would be a good integration test that apparmor works.

LordGrimmauld commented:
fair point in seeing something crash! Will add a test case for `hello` succeeding and `hexdump` failing.

LordGrimmauld commented:
Alright, i managed to add a failing test case for hexdump and a succeeding testcase for hello.

getchoo commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +51,55 @@ in
+                };
+
+                path = lib.mkOption {
+                  type = types.nullOr types.path;
+                  default = null;
+                  description = "A path of a profile file to include. Incompatible with profile.";
```
```suggestion
                  description = "A path of a profile file to include. Incompatible with profile.";
                  type = types.nullOr types.path;
                  default = null;
```
Nitpick for consistency with other declarations

LordGrimmauld commented:
> Since assertions don't really exist in submodules, we can instead create these assertions from the top-level context by mapping over each policy (we can also simplify this assertion with an XOR operator)
@getchoo you are missing one relevant detail here: The current assert runs in the `apply` function. This is relevant, because *querrying* the policy path from outside will always return a path, either taking the path defined in `path` or writing the policy to store and returning that path.
This is relevant, because this means you can't easily check for null/undefined of `<policy>.path` outside.

The alternative is dropping the apply function and inlining it wherever the actual path is required. I did not do this because the previous module using apply. I also figured it'd break peoples configs less hard if behavior stayed roughly similar (previously, you could query the policy path by using `policy.profile` to read).

I am willing to change the assert to a module-global assert instead of specifically in the submodule. I did want to confirm though, because your requested changes alone wouldn't work.

symphorien commented on nixos/tests/apparmor/default.nix:
```diff
@@ -0,0 +1,133 @@
+
+        with subtest("apparmorProfileEnforce"):
+            machine.succeed("${getExe pkgs.hello} 1> /tmp/test-file")
+            machine.fail("${lib.getExe' pkgs.util-linux "hexdump"} /tmp/test-file") # no access to /etc/os-release granted by apparmor
```
```suggestion
            machine.fail("${lib.getExe' pkgs.util-linux "hexdump"} /tmp/test-file") # no access to /tmp/test-file granted by apparmor
```
unless I missed something

LordGrimmauld replied:
Oops, good catch! This is what midnight coding does to a brain, forgot to update the comment!

symphorien commented:
thank you for the added test. I'd like to reread that more carefully when I have more time on my hands, then merge, but if someone wants to merge before that please do.

LordGrimmauld commented:
Currently waiting for getchoos answer whether to prefer module-level asserts or the old apply-function logic because we can't have both. I am happy with either, but wanted confirmation first.
I have resolution of the merge conflicts on my local branch, but waiting to push.

LordGrimmauld commented:
Yeah okay this probably won't work. Seems i can't easily call `isDefined` on the submodule option unless i am missing something.

LordGrimmauld commented:
Alright, while messing with this more, i also noticed the status test is extremely flaky, and i will have to fix that too. While trying to do so, i found a bug in apparmor upstream: https://gitlab.com/apparmor/apparmor/-/issues/470

I Will continue work on this tomorrow.

getchoo commented:
> The alternative is dropping the apply function and inlining it wherever the actual path is required. I did not do this because the previous module using apply.

I think this would end up being the cleaner solution here. Having so much logic in an `apply` is very messy, and in this case interferes with the module system collecting all of the failed assertions of a configuration rather than immediately failing on one

> I also figured it'd break peoples configs less hard if behavior stayed roughly similar (previously, you could query the policy path by using policy.profile to read).

But I'm not really all to sure about this. On one hand it's still possible if you are already defining that option as a path, but it could create issues if people were always relying on it existing. This PR is already introducing some serious breaking changes with the primary `enable` option being removed though, so I wouldn't worry *too* much; just do what you think will improve the API here

> Yeah okay this probably won't work. Seems i can't easily call isDefined on the submodule option unless i am missing something.

This does seem to be the case, but I think I have ~~one of my worst written~~ a very good workaround

```nix
import <nixpkgs/nixos> {
  configuration =
    { config, lib, ... }:
    let
      # Accessing submodule options when not defined results in an error thunk rather than a regular option object
      # We can emulate the behavior of `<option>.isDefined` by attempting to evaluate it instead 
      budgetIsDefined = value: (builtins.tryEval value).success;
    in
    {
      options.myOption = lib.mkOption {
        type = lib.types.attrsOf (
          lib.types.submodule {
            options = {
              # One undefined by default option
              bar = lib.mkOption {
                type = lib.types.str;
              };

              # Nullable option
              baz = lib.mkOption {
                type = lib.types.nullOr lib.types.path;
                default = null;
              };
            };
          }
        );
        default = { };
      };

      config = {
        assertions = lib.mapAttrsToList (name: value: {
          assertion = lib.xor (budgetIsDefined value.bar) (value.baz != null);
          message = "only `myOption.${name}.bar` or `myOption.${name}.baz` may be defined";
        }) config.myOption;

        myOption.foo = { };
      };
    };
}
```

getchoo commented on nixos/modules/security/apparmor.nix:
```diff
@@ -42,22 +57,42 @@ in
-        default = {};
+        type = types.attrsOf (
+          types.submodule (
+            { ... }:
```
```suggestion
```
Nitpick, but this can omitted

getchoo commented on nixos/modules/security/apparmor.nix:
```diff
@@ -90,13 +125,17 @@ in
-      }
-    ) (attrNames cfg.policies);
+      }) (attrNames cfg.policies)
+      ++ lib.mapAttrsToList (policyName: policyCfg: {
+        assertion = lib.xor (policyCfg.path != null) (submoduleOptionIsDefined policyCfg.profile);
+        message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
+      }) cfg.policies;
```
```suggestion
      lib.concatLists (
        lib.mapAttrsToList (policyName: policyCfg: [
          {
            # Because, for instance, aa-remove-unknown uses profiles_names_list() in rc.apparmor.functions
            # which does not recurse into sub-directories.
            assertion = match ".*/.*" policyName == null;
            message = "`security.apparmor.policies.\"${policyName}\"' must not contain a slash.";
          }
          {
            assertion = lib.xor (policyCfg.path != null) (submoduleOptionIsDefined policyCfg.profile);
            message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
          }
        }) cfg.policies
      );
```

Since we're mapping over the same value, it'd be better to define all of the assertions here rather than separately

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/apparmor-on-nixos-roadmap/57217/1

LordGrimmauld commented:
Alright, my new ram arrived, i ran `nixpkgs-review` (which this passed) and fixed the merge conflicts (again). I believe this is ready.

getchoo reviewed (approved):
LGTM. Thanks for all of your work here!

getchoo commented:
@ofborg test apparmor

LordGrimmauld commented:
I am confused, why are ofborg eval and eval-lib-tests failing on gnome circle things? this PR doesn't even touch those... But at least the apparmor nixos test passes, and the new github actions to replace ofborg pass too

mweinelt commented:
They were probably failing on the base commit at the time you last rebased.

LordGrimmauld commented:
i fixed merge conflicts (again).
I also noticed the test being flaky because hexdump wouldn't match the profile properly, fixed by matching against `/nix/store/*/bin/hexdump` for now.

getchoo commented:
@ofborg test apparmor
And feel free to ping me in a couple days if there are no other reviews :)
